src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/QualityGuidelines/CSharpRethrowToPreserveStackDetails.cs

```diff
@@ -62,6 +62,9 @@ private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
                     case SyntaxKind.AnonymousMethodExpression:
                     case SyntaxKind.ClassDeclaration:
                     case SyntaxKind.StructDeclaration:
+#if CODEANALYSIS_V3_OR_BETTER
+                    case SyntaxKind.RecordDeclaration:
```

Comment:
I did a bit of digging and I think we can add a test for the `struct`, `class` and `record` cases doing something like:

```csharp
using System;

public class C
{
    private string? s = default(string) ?? throw new Exception();
}
```

Reply:
This will return very early:

https://github.com/dotnet/roslyn-analyzers/blob/7641932817a0871edc7ba15a22f357a8581ef57c/src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/QualityGuidelines/CSharpRethrowToPreserveStackDetails.cs#L24-L29

It shouldn't get to this code path at all

Reply:
What about this?

```csharp
public class C
{
  private string? S
  {
    get
    {
      throw new Exception();
    }
  }
}
```

Reply:
📝  This is part of the reason why I love code coverage tooling as part of CI builds. If you have code coverage tooling and make an assumption like this for testing, you can see that test cases were eliminated if someone comes along in the future and adds the special cases for accessors or properties.

Reply:
@sharwell That's correct. I was also mistaken of the sample provided by @Evangelink, the expression isn't null. My comment (https://github.com/dotnet/roslyn-analyzers/pull/4232#discussion_r495897805) looks a bit funny.

The suggested tests should improve test coverage (but still there is no behavioral change - it just stops walking up the tree when it's known we can't meet a catch clause).

But I think my suggestion in https://github.com/dotnet/roslyn-analyzers/pull/4232#issuecomment-699975946 should make things a little bit faster in most cases.

The current approach walks up the tree for throw expressions. (the registered action is for SyntaxKind.ThrowStatement).

Registering for CatchStatement instead should be a bit faster for most cases, it also doesn't require this list of switch to return early.

What do you think?

Reply:
@Youssef1313 Having some tests showing nothing is reported is a good test IMO.

Regarding the modification of the analyzer to register on the catch and then walking down, as far as I remember this is a "bad" idea because when you have the callback you don't have any guarantee that what is below the callback has been created. Doing the walk will force the creation but it doesn't go in the normal route. I might be a little wrong on the exact explanation but I remember that going up was ok but down was to avoid.

Reply:
@sharwell I was actually wondering if there was a reason not to have any stop on member declaration because I don't see the need to walk-up anymore. If you agree, I will create a ticket for it.

Reply:
This analyzer can be rewritten as a very simple IOperation analyzer, which will drastically simplify the implementation, make it language agnostic and improve the overall performance. We will also not need these very specific syntax checks. We just walk up the operation tree until we reach a catch operation, or a lambda/local function operation.

Reply:
@mavasani I was waiting for this feedback, I have a PR ready for it :)